parser-typechecker/src/Unison/Codebase/Causal/Type.hs

```diff
@@ -0,0 +1,139 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+module Unison.Codebase.Causal.Type
+  ( Causal (..),
+    RawHash (..),
+    pattern One,
+    pattern Cons,
+    pattern Merge,
+    before,
+    children,
+    lca,
+    head_,
+  )
+where
+
+import qualified Control.Lens as Lens
+import qualified Data.Map as Map
+import qualified Data.Sequence as Seq
+import qualified Data.Set as Set
+import Unison.Hash (Hash)
+import Unison.Prelude
+import Prelude hiding (head, read, tail)
+
+{-
+`Causal a` has 5 operations, specified algebraically here:
+
+* `before : Causal m a -> Causal m a -> m Bool` defines a partial order on
+            `Causal`.
+* `head : Causal m a -> a`, which represents the "latest" `a` value in a causal
+          chain.
+* `one : a -> Causal m a`, satisfying `head (one hd) == hd`
+* `cons : a -> Causal a -> Causal a`, satisfying `head (cons hd tl) == hd` and
+          also `before tl (cons hd tl)`.
+* `merge : CommutativeSemigroup a => Causal a -> Causal a -> Causal a`, which is
```

Comment:
Not sure what this is implying; The fact that this is a type signature implies the existence of a `CommutativeSemigroup` class somewhere which I can't find in this PR. I suspect you mean it more conceptually, e.g. "You need to tell me how to merge two or more branches in order to merge", but I think I'd rather have that written in English rather than referring to classes we don't actually use.

I understand that it's _kind of_ true,